cmd/csi-api-gen/generators/api.go

```
@@ -0,0 +1,432 @@
+// here we actually directly process versions' packages directly.
+func buildCanonicalAPIGroupDefinition(pkgPath string, pkg *types.Package, groups map[string]*groupDefinition) {
+	groupNameAndVersion := strings.TrimPrefix(pkgPath, csiProxyAPIPath)
+	parts := strings.Split(groupNameAndVersion, "/")
+
+	if len(parts) == 1 {
+		// means it's the group's root directory, no need to process it at this time,
+		// we'll get around to it when we process its versions' packages
+		return
+	}
+
+	if len(parts) != 2 || !apiversion.IsValidVersion(parts[1]) {
+		klog.Fatalf("Unexpected go package %q, should be of the form \"%s<api_group_name>/<version>\", where <version> should be a valid API group version identifier",
+			pkgPath, csiProxyAPIPath)
+	}
+
+	groupPath := csiProxyAPIPath + parts[0]
+	definition, present := groups[groupPath]
+	if !present {
+		klog.V(5).Infof("Found API group %q", parts[0])
+		definition = newGroupDefinition(parts[0], csiProxyAPIPath+parts[0])
+		groups[groupPath] = definition
+	}
+	definition.addVersion(pkg)
+
+	klog.V(5).Infof("Found version %q for API group %q", parts[1], parts[0])
+}
+
+// removeGeneratedFiles removes all generated files before we re-generate them anew.
+func removeGeneratedFiles(group *groupDefinition, outputBase string) {
+	toRemove := []string{
+		path.Join(group.serverPkg(), "api_group_generated.go"),
+		path.Join(group.internalServerPkg(), "types_generated.go"),
+	}
+	for _, version := range group.versions {
+		toRemove = append(toRemove,
+			path.Join(group.versionedServerPkg(version.Name), "conversion_generated.go"),
+			path.Join(group.versionedServerPkg(version.Name), "server_generated.go"),
+			path.Join(group.versionedClientPkg(version.Name), "client_generated.go"))
+	}
+
+	for _, filePath := range toRemove {
+		filePath = path.Join(outputBase, filePath)
+		if err := os.RemoveAll(filePath); err != nil {
+			klog.Fatalf("unable to remove %q: %v", filePath, err)
```

do we want to retry or just fail immediately?

---

I'm not sure what error modes could justify retrying later?

Also, the general philosophy here (and also for all of k8s' code generation, as far as I can tell) is that since this is not a service, and is just meant to be run manually (or in tests) from time to time, it's okay to fail early.